core/templates/components/state-directives/outcome-editor/outcome-feedback-editor.component.spec.ts

```diff
@@ -0,0 +1,96 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for OutcomeFeedbackEditorComponent.
+ */
+
+import { ChangeDetectorRef, NO_ERRORS_SCHEMA } from '@angular/core';
+import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
+import { ContextService } from 'services/context.service';
+import { OutcomeFeedbackEditorComponent } from './outcome-feedback-editor.component';
+
+describe('Outcome Feedback Editor Component', () => {
+  let fixture: ComponentFixture<OutcomeFeedbackEditorComponent>;
+  let component: OutcomeFeedbackEditorComponent;
+  let contextService: ContextService;
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [],
+      declarations: [
+        OutcomeFeedbackEditorComponent
+      ],
+      providers: [
+        ChangeDetectorRef,
+        ContextService
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(OutcomeFeedbackEditorComponent);
+    component = fixture.componentInstance;
+    contextService = TestBed.inject(ContextService);
+  });
+
+  it('should set outcome feedback schema on initialization', () => {
+    expect(component.OUTCOME_FEEDBACK_SCHEMA).toBe(undefined);
+
+    spyOn(contextService, 'getEntityType').and.returnValue('notQuestion');
+    component.ngOnInit();
+
+    expect(component.OUTCOME_FEEDBACK_SCHEMA).toEqual({
+      type: 'html',
+      ui_config: {
+        hide_complex_extensions: false
+      }
+    });
+  });
+
+  it('should get schema', () => {
+    spyOn(contextService, 'getEntityType').and.returnValue('question');
+    component.ngOnInit();
+    expect(component.getSchema()).toEqual({
+      type: 'html',
+      ui_config: {
+        hide_complex_extensions: true
+      }
+    });
+  });
+
+  it('should update html', () => {
+    component.outcome = {
+      feedback: {
+        html: '<p> Previous HTML string </p>'
+      },
+      hasNonemptyFeedback: () => true
+    };
+
+    component.updateHtml('<p> New HTML string </p>');
+
+    expect(component.outcome.feedback.html).toBe('<p> New HTML string </p>');
+  });
+
+  it('should not update html if new html string is same as old', () => {
+    component.outcome = {
+      feedback: {
+        html: '<p> Previous HTML string </p>'
+      },
+      hasNonemptyFeedback: () => true
+    };
+    component.updateHtml('<p> Previous HTML string </p>');
+  });
```

Comment:
This test has no expectation and therefore doesn't test what its description claims it is testing.

Reply:
Hmm, I don't think we would be able to check if changeDetectorRef was called using a spy and `expect(component.outcome.feedback.html).toBe(<p> Previous HTML string </p>');` won't make sense. 

Removed the test

Reply:
Hm I don't see at first glance why this is untestable. Could you spy on this.outcome.feedback.html to make sure it does not get changed (or on changeDetectorRef for that matter)?

Reply:
To spy on onChangeDetector we'll have to do: `(component as any).changeDettectorRef, 'detectChanges()'` and as using `any` is not allowed, so we can't do it (unless we use a lint disable line).

For `this.outcome.feedback.html`, it already has a value (e.g. `Old string`), and we update it with `Old string`, then we can't confirm if it was updated or not by spying on `this.outcome.feedback.html`, as it would be `Old string` in both cases (updated or not).

Reply:
You can always spy on prototypes, right? I did this for subject once. https://github.com/oppia/oppia/blob/develop/core/templates/app-events/event-bus.service.spec.ts#L56

Maybe this could be one test?
```
const changeDetectorRef = fixture.debugElement.injector.get(
      ChangeDetectorRef);
    const detectChangesSpy = spyOn(
      changeDetectorRef.constructor.prototype, 'detectChanges');
    component.outcome = {
      feedback: {
        html: '<p> Previous HTML string </p>'
      },
      hasNonemptyFeedback: () => true
    };

    expect(component.outcome.feedback.html).toBe(
      '<p> Previous HTML string </p>');

    component.updateHtml('<p> Previous HTML string </p>');
    expect(detectChangesSpy).not.toHaveBeenCalled();
```